core/src/main/java/org/apache/iceberg/BaseIncrementalChangelogScan.java

```diff
@@ -63,33 +60,43 @@ protected CloseableIterable<ChangelogScanTask> doPlanFiles(
       return CloseableIterable.empty();
     }
 
-    Set<Long> changelogSnapshotIds = toSnapshotIds(changelogSnapshots);
+    Map<Long, Integer> snapshotOrdinals = computeSnapshotOrdinals(changelogSnapshots);
 
-    Set<ManifestFile> newDataManifests =
-        FluentIterable.from(changelogSnapshots)
-            .transformAndConcat(snapshot -> snapshot.dataManifests(table().io()))
-            .filter(manifest -> changelogSnapshotIds.contains(manifest.snapshotId()))
-            .toSet();
-
-    ManifestGroup manifestGroup =
-        new ManifestGroup(table().io(), newDataManifests, ImmutableList.of())
-            .specsById(table().specs())
-            .caseSensitive(isCaseSensitive())
-            .select(scanColumns())
-            .filterData(filter())
-            .filterManifestEntries(entry -> changelogSnapshotIds.contains(entry.snapshotId()))
-            .ignoreExisting()
-            .columnsToKeepStats(columnsToKeepStats());
-
-    if (shouldIgnoreResiduals()) {
-      manifestGroup = manifestGroup.ignoreResiduals();
-    }
-
-    if (newDataManifests.size() > 1 && shouldPlanWithExecutor()) {
-      manifestGroup = manifestGroup.planWith(planExecutor());
-    }
+    // map of delete file to the snapshot where the delete file is added
+    // the delete file is keyed by its path, and the snapshot is represented by the snapshot ordinal
+    Map<String, Integer> deleteFileToSnapshotOrdinal =
+        computeDeleteFileToSnapshotOrdinal(changelogSnapshots, snapshotOrdinals);
 
-    return manifestGroup.plan(new CreateDataFileChangeTasks(changelogSnapshots));
+    Iterable<CloseableIterable<ChangelogScanTask>> plans =
+        FluentIterable.from(changelogSnapshots)
+            .transform(
+                snapshot -> {
+                  List<ManifestFile> dataManifests = snapshot.dataManifests(table().io());
+                  List<ManifestFile> deleteManifests = snapshot.deleteManifests(table().io());
+
+                  ManifestGroup manifestGroup =
+                      new ManifestGroup(table().io(), dataManifests, deleteManifests)
+                          .specsById(table().specs())
+                          .caseSensitive(isCaseSensitive())
+                          .select(scanColumns())
+                          .filterData(filter())
+                          .columnsToKeepStats(columnsToKeepStats());
+
+                  if (shouldIgnoreResiduals()) {
+                    manifestGroup = manifestGroup.ignoreResiduals();
+                  }
+
+                  if (dataManifests.size() > 1 && shouldPlanWithExecutor()) {
+                    manifestGroup = manifestGroup.planWith(planExecutor());
+                  }
+
+                  long snapshotId = snapshot.snapshotId();
+                  return manifestGroup.plan(
+                      new CreateDataFileChangeTasks(
+                          snapshotId, snapshotOrdinals, deleteFileToSnapshotOrdinal));
```

Comment:
I'm not sure that we can filter with the current snapshotId. I think we need to filter with the `toSnapshotId` of the scan, so we don't emit records which are added and deleted when the `IncrementalScan` is for multiple snapshots.

I have asked the corresponding question on the dev list thread, but for reference:

> What is the expected behaviour when the `IncrementalScan` is created for not a single snapshot, but for multiple snapshots?
> - S1 added PK1-V1
> - S2 updated PK1-V1 to PK1-V1b (removed PK1-V1 and added PK1-V1b)
> - S3 updated PK1-V1b to PK1-V1c (removed PK1-V1b and added PK1-V1c)
> 
> Let's say we have IncrementalScan.fromSnapshotInclusive(S1).toSnapshot(S3).
> Or we need to return:
> (a)
> - PK1,V1c,INSERTED
> 
> Or is it ok, to return:
> (b)
> - PK1,V1,INSERTED
> - PK1,V1,DELETED
> - PK1,V1b,INSERTED
> - PK1,V1b,DELETED
> - PK1,V1c,INSERTED

I think the (a) is the correct behaviour.

Reply:
Ah, in this case, (b) is the correct behavior. The changelog scan is an incremental scan over multiple snapshots in a range, and should emit the changes for each snapshot. This is the current behavior for the supported case, which is copy-on-write. What you are seeking are the net changes, which is functionality that is also supported by Spark, and built on top of the changelog scan. This uses `ChangelogIterator.removeNetCarryovers`. This functionality is exposed in the Spark procedure, `create_changelog_view`. (Of course, one can also use it programmatically.)

Reply:
As agreed on the dev list, let's proceed with the snapshot-by-snapshot reader for now, and decide on the net_changes solution later